### PR TITLE
[PPR Navs] Bugfix: Dynamic data never streams in if prefetch entry is stale

### PR DESCRIPTION
**Pull request by acdlite**
Adds a regression test and a fix for a bug that sometimes happens when a prefetched route on the client becomes stale — the app would get stuck in a loading state.

The problem was the condition I used to fallback to the non-PPR implementation, inside navigateReducer. It was too narrow, causing prefetched segments that contained dynamic holes to sometimes be treated as if they were complete. The net effect was that the dynamic data would never stream in, and the page would get stuck in a fallback state until the stale prefetch was eventually purged from the cache, or the user refreshed the page.

The reason the mistake happened was, as an incremental step, I decided to fallback to the non-PPR implementation for any case where I hadn't yet implemented the equivalent functionality. I think still think this is a good strategy, despite the mistake, but I'm eager to get everything migrated to the new model as soon as possible.

Closes NEXT-1920

### CODE REVIEW

File: test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts [hunk 1]
```diff
@@ -94,6 +90,88 @@ describe('ppr-navigations', () => {
...
+            super(__FAKE_NOW__)
+          } else {
+            super(...args)
+          }
+        }
+        static now() {
+          return __FAKE_NOW__
+        }
+      }
+    `)
```
Comment (acdlite) [inline test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts:144]: This was the least bad way I could find to simulate time elapsing. Turns out this is a known gap in Playwright: https://github.com/microsoft/playwright/issues/6347

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong></summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| buildDuration | 12.8s | 12.8s | N/A |
| buildDurationCached | 7.1s | 6.2s | N/A |
| nodeModulesSize | 200 MB | 200 MB | ⚠️ +4.95 kB |
| nextStartRea..uration (ms) | 425ms | 430ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| 193.HASH.js gzip | 181 B | 182 B | N/A |
| 3f784ff6-HASH.js gzip | 53.3 kB | 53.3 kB | ✓ |
| 433-HASH.js gzip | 28.2 kB | 28.3 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 239 B | 242 B | N/A |
| main-HASH.js gzip | 31.7 kB | 31.7 kB | N/A |
| webpack-HASH.js gzip | 1.7 kB | 1.7 kB | N/A |
| Overall change | 98.5 kB | 98.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| _app-HASH.js gzip | 194 B | 195 B | N/A |
| _error-HASH.js gzip | 183 B | 181 B | N/A |
| amp-HASH.js gzip | 504 B | 502 B | N/A |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 255 B | 253 B | N/A |
| head-HASH.js gzip | 350 B | 349 B | N/A |
| hooks-HASH.js gzip | 369 B | 369 B | ✓ |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 255 B | 256 B | N/A |
| link-HASH.js gzip | 2.61 kB | 2.61 kB | ✓ |
| routerDirect..HASH.js gzip | 312 B | 311 B | N/A |
| script-HASH.js gzip | 385 B | 383 B | N/A |
| withRouter-HASH.js gzip | 307 B | 308 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 3.4 kB | 3.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| index.html gzip | 528 B | 527 B | N/A |
| link.html gzip | 540 B | 539 B | N/A |
| withRouter.html gzip | 523 B | 523 B | ✓ |
| Overall change | 523 B | 523 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| edge-ssr.js gzip | 93.7 kB | 93.7 kB | N/A |
| page.js gzip | 147 kB | 147 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 624 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 37.4 kB | 37.4 kB | N/A |
| edge-runtime..pack.js gzip | 1.92 kB | 1.92 kB | ✓ |
| Overall change | 2.07 kB | 2.07 kB | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | acdlite/next.js fix-dynamic-data-never-streams | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 168 kB | 168 kB | ✓ |
| app-page-exp..prod.js gzip | 94.2 kB | 94.2 kB | ✓ |
| app-page-tur..prod.js gzip | 94.9 kB | 94.9 kB | ✓ |
| app-page-tur..prod.js gzip | 89.4 kB | 89.4 kB | ✓ |
| app-page.run...dev.js gzip | 138 kB | 138 kB | ✓ |
| app-page.run..prod.js gzip | 88.7 kB | 88.7 kB | ✓ |
| app-route-ex...dev.js gzip | 24.1 kB | 24.1 kB | ✓ |
| app-route-ex..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.7 kB | 16.7 kB | ✓ |
| app-route-tu..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| app-route.ru...dev.js gzip | 23.5 kB | 23.5 kB | ✓ |
| app-route.ru..prod.js gzip | 16.3 kB | 16.3 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.64 kB | 9.64 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.37 kB | 9.37 kB | ✓ |
| pages-turbo...prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| pages.runtim...dev.js gzip | 22.5 kB | 22.5 kB | ✓ |
| pages.runtim..prod.js gzip | 21.9 kB | 21.9 kB | ✓ |
| server.runti..prod.js gzip | 49.4 kB | 49.4 kB | ✓ |
| Overall change | 932 kB | 932 kB | ✓ |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>page.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>433-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 215e49725ad2640f059d59292c2534b32118a191

File: packages/next/src/client/components/router-reducer/reducers/navigate-reducer.ts [hunk 2]
```diff
@@ -413,22 +413,32 @@ function navigateReducer_PPR(
...
+              prefetchEntryCacheStatus === PrefetchCacheEntryStatus.stale
+
             const task = updateCacheNodeOnNavigation(
               currentCache,
               currentTree,
               prefetchedTree,
               seedData,
-              head
+              head,
+              isPrefetchStale
```
Comment (acdlite) [inline packages/next/src/client/components/router-reducer/reducers/navigate-reducer.ts:441]: The fix is that I moved the stale check into this branch and reimplemented the `addRefetchToLeafSegments` logic inside `updateCacheNodeOnNavigation`. The idea is that we shouldn't show the prefetch data, even temporarily, if it's too stale.

Comment (acdlite) [reply packages/next/src/client/components/router-reducer/reducers/navigate-reducer.ts:441]: Or at least that makes it match the current Next.js behavior, not sure it's ideal longer term.

File: test/lib/browsers/playwright.ts [hunk 3]
```diff
@@ -468,6 +468,10 @@ export class Playwright extends BrowserInterface {
     return page.evaluate<T>(fn).catch(() => null)
   }
 
+  getPlayrightPage() {
```
Comment (feedthejim) [inline test/lib/browsers/playwright.ts:471]: typo

File: test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts (hunk 1 above)
Comment (feedthejim) [reply test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts:144]: did you the logic I added for the client-cache tests as well?

Comment (feedthejim) [reply test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts:144]: kinda doing the same things

Comment (acdlite) [reply test/e2e/app-dir/ppr-navigations/ppr-navigations.test.ts]: Oh no I didn't but I just looked it up, that's neat that we independently used the same pattern.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 05f43c3a433052d8dee44c92231553631559b78d

`pnpm test-dev test/development/typescript-auto-install/index.test.ts` 
- typescript-auto-install > should detect TypeScript being added and auto setup

<details>
<summary>Expand output</summary>

  ● typescript-auto-install › should detect TypeScript being added and auto setup

    TIMED OUT: /hello again/

    <head><meta charset="utf-8"><meta name="viewport" content="width=device-width"><meta name="next-head-count" content="2"><noscript data-n-css=""></noscript><script src="/_next/static/chunks/35c6e_next_df05dd._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/c8460_react_22c55a._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/4e99c_react-dom_cjs_react-dom_development_c0b264.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/4e99c_react-dom_55c180._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/node_modules__pnpm_8959e2._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Broot%20of%20the%20server%5D__272bc4._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/35c6e_next_dist_pages_283a5a._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_a42a90._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_62e145._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/pages/_app.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/35c6e_next_dist_c7b83a._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/c8460_react_3cd7ff._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Broot%20of%20the%20server%5D__94958f._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/35c6e_next_dist_pages_15a886._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_082619._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_a1df65._.js?ts=1703169362135" defer=""></script><script src="/_next/static/chunks/pages/index.js?ts=1703169362135" defer=""></script><script src="/_next/static/development/_ssgManifest.js?ts=1703169362135" defer=""></script><script src="/_next/static/development/_buildManifest.js?ts=1703169362135" defer=""></script><noscript id="__next_css__DO_NOT_USE__"></noscript></head><body style="overflow: hidden;"><div id="__next"><p>hello world</p></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{}},"page":"/","query":{},"buildId":"development","nextExport":true,"autoExport":true,"isFallback":false,"scriptLoader":[]}</script><script src="/_next/static/chunks/%5Broot%20of%20the%20server%5D__65858d._.js"></script><script src="/_next/static/chunks/35c6e_next_dist_pages_7ec749._.js"></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_87555b._.js"></script><script src="/_next/static/chunks/%5Bnext%5D_entry_page-loader_ts_cf7c9b._.js"></script><div id="__next-build-watcher" style="position: fixed; bottom: 10px; right: 20px; width: 0px; height: 0px; z-index: 99999;"></div><next-route-announcer><p aria-live="assertive" id="__next-route-announcer__" role="alert" style="border: 0px; clip: rect(0px, 0px, 0px, 0px); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; top: 0px; width: 1px; white-space: nowrap; overflow-wrap: normal;"></p></next-route-announcer><nextjs-portal></nextjs-portal></body>

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (development/typescript-auto-install/index.test.ts:63:5)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Review by ztanner — APPROVED** [review]
nice catch

**Pull request merged**